File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/holidayisland/domain/HolidayIslandGameState.java
package com.zwstudio.logicpuzzlesandroid.puzzles.holidayisland.domain;

import com.zwstudio.logicpuzzlesandroid.common.domain.AllowedObjectState;
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGameState;
import com.zwstudio.logicpuzzlesandroid.common.domain.Graph;
import com.zwstudio.logicpuzzlesandroid.common.domain.MarkerOptions;
import com.zwstudio.logicpuzzlesandroid.common.domain.Node;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState;
import com.zwstudio.logicpuzzlesandroid.puzzles.tierradelfuego.domain.TierraDelFuegoGame;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import fj.F;

import static fj.data.HashMap.fromMap;

public class HolidayIslandGameState extends CellsGameState<HolidayIslandGame, HolidayIslandGameMove, HolidayIslandGameState> {
    public HolidayIslandObject[] objArray;

    public HolidayIslandGameState(HolidayIslandGame game) {
        super(game);
        objArray = new HolidayIslandObject[rows() * cols()];
        for (int i = 0; i < objArray.length; i++)
            objArray[i] = new HolidayIslandEmptyObject();
        for (Map.Entry<Position, Integer> entry : game.pos2hint.entrySet()) {
            Position p = entry.getKey();
            int n = entry.getValue();
            set(p, new HolidayIslandHintObject() {{tiles = n;}});
        }
    }

    public HolidayIslandObject get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public HolidayIslandObject get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, HolidayIslandObject dotObj) {
        objArray[row * cols() + col] = dotObj;
    }
    public void set(Position p, HolidayIslandObject obj) {
        set(p.row, p.col, obj);
    }

    public boolean setObject(HolidayIslandGameMove move) {
        if (!isValid(move.p) || game.pos2hint.get(move.p) != null || get(move.p).equals(move.obj)) return false;
        set(move.p, move.obj);
        updateIsSolved();
        return true;
    }

    public boolean switchObject(HolidayIslandGameMove move) {
        if (!isValid(move.p) || game.pos2hint.get(move.p) != null) return false;
        MarkerOptions markerOption = MarkerOptions.values()[game.gdi.getMarkerOption()];
        F<HolidayIslandObject, HolidayIslandObject> f = obj -> {
            if (obj instanceof HolidayIslandEmptyObject)
                return markerOption == MarkerOptions.MarkerFirst ?
                        new HolidayIslandMarkerObject() : new HolidayIslandTreeObject();
            if (obj instanceof HolidayIslandTreeObject)
                return markerOption == MarkerOptions.MarkerLast ?
                        new HolidayIslandMarkerObject() : new HolidayIslandEmptyObject();
            if (obj instanceof HolidayIslandMarkerObject)
                return markerOption == MarkerOptions.MarkerFirst ?
                        new HolidayIslandTreeObject() : new HolidayIslandEmptyObject();
            return obj;
        };
        HolidayIslandObject o = get(move.p);
        move.obj = f.f(o);
        return setObject(move);
    }

    /*
        iOS Game: Logic Games/Puzzle Set 11/Holiday Island

        Summary
        This time the campers won't have their way!

        Description
        1. This time the resort is an island, the place is packed and the campers
           (Tents) must compromise!
        2. The board represents an Island, where there are a few Tents, identified
           by the numbers.
        3. Your job is to find the water surrounding the island, with these rules:
        4. There is only one, continuous island.
        5. The numbers tell you how many tiles that camper can walk from his Tent,
           by moving horizontally or vertically. A camper can't cross water or
           other Tents.
    */
    private void updateIsSolved() {
        boolean allowedObjectsOnly = game.gdi.isAllowedObjectsOnly();
        isSolved = true;
        List<Position> rngHints = new ArrayList<>();
        Graph g = new Graph();
        Map<Position, Node> pos2node = new HashMap<>();
        for (int r = 0; r < rows(); r++)
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                HolidayIslandObject o = get(p);
                if (o instanceof HolidayIslandForbiddenObject)
                    set(p, new HolidayIslandEmptyObject());
                else if (o instanceof HolidayIslandTreeObject)
                    ((HolidayIslandTreeObject)o).state = AllowedObjectState.Normal;
                else if (o instanceof HolidayIslandHintObject) {
                    ((HolidayIslandHintObject)o).state = HintState.Normal;
                    rngHints.add(p);
                }
                if (!(o instanceof HolidayIslandTreeObject)) {
                    Node node = new Node(p.toString());
                    g.addNode(node);
                    pos2node.put(p, node);
                }
            }
        for (Map.Entry<Position, Node> entry : pos2node.entrySet()) {
            Position p = entry.getKey();
            Node node = entry.getValue();
            for (Position os : TierraDelFuegoGame.offset) {
                Position p2 = p.add(os);
                Node node2 = pos2node.get(p2);
                if (node2 != null) g.connectNode(node, node2);
            }
        }
        {
            // 4. There is only one, continuous island.
            g.setRootNode(fromMap(pos2node).values().head());
            List<Node> nodeList = g.bfs();
            if (nodeList.size() != pos2node.size()) isSolved = false;
        }
        g = new Graph();
        pos2node.clear();
        for (int r = 0; r < rows(); r++)
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                HolidayIslandObject o = get(p);
                if (!(o instanceof HolidayIslandTreeObject || o instanceof HolidayIslandHintObject)) {
                    // 5. A camper can't cross water or other Tents.
                    Node node = new Node(p.toString());
                    g.addNode(node);
                    pos2node.put(p, node);
                }
            }
        for (Map.Entry<Position, Node> entry : pos2node.entrySet()) {
            Position p = entry.getKey();
            Node node = entry.getValue();
            for (Position os : TierraDelFuegoGame.offset) {
                Position p2 = p.add(os);
                Node node2 = pos2node.get(p2);
                if (node2 == null) continue;
                g.connectNode(node, node2);
            }
        }
        List<List<Position>> areas = new ArrayList<>();
        Map<Position, Integer> pos2area = new HashMap<>();
        while (!pos2node.isEmpty()) {
            g.setRootNode(fromMap(pos2node).values().head());
            List<Node> nodeList = g.bfs();
            List<Position> area = fromMap(pos2node).toStream().filter(e -> nodeList.contains(e._2())).map(e -> e._1()).toJavaList();
            int n = areas.size();
            for (Node node : nodeList) {
                Position p = fromMap(pos2node).toStream().find(e -> e._2().equals(node)).some()._1();
                pos2node.remove(p);
                pos2area.put(p, n);
            }
            areas.add(area);
        }
        for (Position p : rngHints) {
            int n2 = game.pos2hint.get(p);
            Set<Position> rng = new HashSet<>();
            for (Position os : HolidayIslandGame.offset) {
                Position p2 = p.add(os);
                Integer i = pos2area.get(p2);
                if (i == null) continue;
                rng.addAll(areas.get(i));
            }
            int n1 = rng.size();
            // 5. The numbers tell you how many tiles that camper can walk from his Tent,
            // by moving horizontally or vertically.
            HintState s = n1 > n2 ? HintState.Normal : n1 == n2 ? HintState.Complete : HintState.Error;
            ((HolidayIslandHintObject)get(p)).state = s;
            if (s != HintState.Complete) isSolved = false;
            if (allowedObjectsOnly && n1 <= n2)
                for(Position p2 : rng)
                    if (!p2.equals(p))
                        set(p2, new HolidayIslandForbiddenObject());
        }
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/holidayisland/domain/HolidayIslandGameState.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.holidayisland.domain

import com.zwstudio.logicpuzzlesandroid.common.domain.*
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState

class HolidayIslandGameState(game: HolidayIslandGame) : CellsGameState<HolidayIslandGame, HolidayIslandGameMove, HolidayIslandGameState>(game) {
    var objArray = Array<HolidayIslandObject>(rows * cols) {  HolidayIslandEmptyObject() }

    operator fun get(row: Int, col: Int) = objArray[row * cols + col]
    operator fun get(p: Position) = this[p.row, p.col]
    operator fun set(row: Int, col: Int, dotObj: HolidayIslandObject) {objArray[row * cols + col] = dotObj}
    operator fun set(p: Position, obj: HolidayIslandObject) {this[p.row, p.col] = obj}

    init {
        for ((p, n) in game.pos2hint)
            this[p] = HolidayIslandHintObject(tiles = n)
    }

    fun setObject(move: HolidayIslandGameMove): Boolean {
        if (!isValid(move.p) || game.pos2hint[move.p] != null || this[move.p] == move.obj) return false
        this[move.p] = move.obj
        updateIsSolved()
        return true
    }

    fun switchObject(move: HolidayIslandGameMove): Boolean {
        if (!isValid(move.p) || game.pos2hint[move.p] != null) return false
        val markerOption = MarkerOptions.values()[game.gdi.markerOption]
        val o = this[move.p]
        move.obj = when(o) {
            is HolidayIslandEmptyObject -> if (markerOption == MarkerOptions.MarkerFirst) HolidayIslandMarkerObject() else HolidayIslandTreeObject()
            is HolidayIslandTreeObject -> if (markerOption == MarkerOptions.MarkerLast) HolidayIslandMarkerObject() else HolidayIslandEmptyObject()
            is HolidayIslandMarkerObject -> if (markerOption == MarkerOptions.MarkerFirst) HolidayIslandTreeObject() else HolidayIslandEmptyObject()
            else -> o
        }
        return setObject(move)
    }

    /*
        iOS Game: Logic Games/Puzzle Set 11/Holiday Island

        Summary
        This time the campers won't have their way!

        Description
        1. This time the resort is an island, the place is packed and the campers
           (Tents) must compromise!
        2. The board represents an Island, where there are a few Tents, identified
           by the numbers.
        3. Your job is to find the water surrounding the island, with these rules:
        4. There is only one, continuous island.
        5. The numbers tell you how many tiles that camper can walk from his Tent,
           by moving horizontally or vertically. A camper can't cross water or
           other Tents.
    */
    private fun updateIsSolved() {
        val allowedObjectsOnly = game.gdi.isAllowedObjectsOnly
        isSolved = true
        val rngHints = mutableListOf<Position>()
        var g = Graph()
        val pos2node = mutableMapOf<Position, Node>()
        for (r in 0 until rows)
            for (c in 0 until cols) {
                val p = Position(r, c)
                val o = this[p]
                if (o is HolidayIslandForbiddenObject)
                    this[p] = HolidayIslandEmptyObject()
                else if (o is HolidayIslandHintObject) {
                    o.state = HintState.Normal
                    rngHints.add(p)
                }
                if (o !is HolidayIslandTreeObject) {
                    val node = Node(p.toString())
                    g.addNode(node)
                    pos2node[p] = node
                }
            }
        for ((p, node) in pos2node) {
            for (os in HolidayIslandGame.offset) {
                val p2 = p.add(os)
                val node2 = pos2node[p2]
                if (node2 != null)
                    g.connectNode(node, node2)
            }
        }
        run {

            // 4. There is only one, continuous island.
            g.rootNode = pos2node.values.first()
            val nodeList = g.bfs()
            if (nodeList.size != pos2node.size) isSolved = false
        }
        g = Graph()
        pos2node.clear()
        for (r in 0 until rows)
            for (c in 0 until cols) {
                val p = Position(r, c)
                val o = get(p)
                if (!(o is HolidayIslandTreeObject || o is HolidayIslandHintObject)) {
                    // 5. A camper can't cross water or other Tents.
                    val node = Node(p.toString())
                    g.addNode(node)
                    pos2node[p] = node
                }
            }
        for ((p, node) in pos2node) {
            for (os in HolidayIslandGame.offset) {
                val p2 = p.add(os)
                val node2 = pos2node[p2] ?: continue
                g.connectNode(node, node2)
            }
        }
        val areas = mutableListOf<List<Position>>()
        val pos2area = mutableMapOf<Position, Int>()
        while (pos2node.isNotEmpty()) {
            g.rootNode = pos2node.values.first()
            val nodeList = g.bfs()
            val area = pos2node.filter { nodeList.contains(it.value) }.map { it.key }
            val n = areas.size
            for (node in nodeList) {
                val p = pos2node.filter { it.value == node }.keys.first()
                pos2node.remove(p)
                pos2area[p] = n
            }
            areas.add(area)
        }
        for (p in rngHints) {
            val n2 = game.pos2hint[p]!!
            val rng = mutableSetOf<Position>()
            for (os in HolidayIslandGame.offset) {
                val p2 = p.add(os)
                val i = pos2area[p2] ?: continue
                rng.addAll(areas[i])
            }
            val n1 = rng.size
            // 5. The numbers tell you how many tiles that camper can walk from his Tent,
            // by moving horizontally or vertically.
            val s = if (n1 > n2) HintState.Normal else if (n1 == n2) HintState.Complete else HintState.Error
            (this[p] as HolidayIslandHintObject).state = s
            if (s != HintState.Complete) isSolved = false
            if (allowedObjectsOnly && n1 <= n2)
                for (p2 in rng)
                    if (p2 != p)
                        this[p2] = HolidayIslandForbiddenObject()
        }
    }
}